### ISSUE TITLE
Allow user defined functions in formulas

### ISSUE THREAD

**Issue by assignUser** [kgoldfeld/simstudy#14]
Hey!

While using `simstudy` I needed certain data that were dependent on variables previously  defined through `defData` but not easily translatable to one of the default distributions. I solved this "outside" of simstudy with a function and `mapply`. 

As this introduces additional elements in the script and breaks definition tables as a complete overview of the defined data, I feel that it would be nice to extend the definition tables/arguments to make that possible. 

I have only had a glimpse at the responsible code, so I am not sure how much work this would be. But if you are open to the idea I would like to try. Maybe functions can be added to definition tables to have all needed information in one place? Let me know what you think! 😃

**Comment by kgoldfeld** [comment]
I am open to the idea of expanding the data definition tables, but would need to understand better what the goal is. My original goal was to keep the table as simple as possible without limiting flexibility, but if you have new ideas about how to increase the range of possible simulations with the addition of a field or two, I am game.

**Comment by assignUser** [comment]
A very simple example to clarify what I mean.
We have this data definition:
```R
def <- dataDef(varname = "index", dist = "pseudorandomSeq", formula = "1+2+3+4+5;5" link = "fill")
```
We have a vector of intensities (or what ever experimental parameter you want it to be) and want the data to have a col "intensity" that contains the `intensities[index]` which is, afaik, not possible with simstudy because only variables defined in the data will work.

Currently we need to do this:
```R
d <- genData(25,def)
d$intensity <- intensities[d$index]
```
Which works and could be solved "inside" simstudy using a whole lot of `defCondition`. 
But maybe there is a way to introduce functions/objects into the definition. Maybe via a new "distribution" with a different formula check in  .devEval (->#13 ). First idea: `formula` gets the function object and arguments are passed via `variance` or `link` ?

**Comment by kgoldfeld** [comment]
I've taken a look. I really like what you've done with the documentation - really nice. And while I haven't checked out the unit tests, that is also a great addition - I've been meaning to get do that, but haven't found the time.

I checked out the new pseudo distributions. In simstudy it is sort of possible to do the pseudorandom distribution using assignTrt, but I have always found that unsatisfying and not sure why I never got around to building a function precisely like you've created. So, that will be a good addition. I'm not sure I like the name of the distribution - we can think about that.

As for pseudorandomSeq, I don't totally get what the goal is - my guess it is not doing exactly what you want it to do. The reason I think this is if you use 

`formula = "1+2+3+4;3"`

or

`formula = "1+2+3+4;4"`

or

`formula = "1+2+3+4;5"`

with pseudorandomSeq, you get the same results.

```
def <- defData(varname = "x", dist = "nonrandom", formula = "1")
def <- defData(def, varname = "xPseudoSeq", dist = "pseudorandomSeq", 
               formula = "1+2+3+4;3" , 
               link = "fill")

set.seed(112)
dx <- genData(100, def)

matrix(dx[, xPseudoSeq], nrow = 4)[,1:5]

###

def <- updateDef(def, "xPseudoSeq", newformula = "1+2+3+4;4")

set.seed(112)
dx <- genData(100, def)

matrix(dx[, xPseudoSeq], nrow = 4)[,1:5]

###

def <- updateDef(def, "xPseudoSeq", newformula = "1+2+3+4;5")

set.seed(112)
dx <- genData(100, def)

matrix(dx[, xPseudoSeq], nrow = 4)[,1:5]
```

Finally, is there any reason we couldn't just specificy the argument as

`formula = "4;3"`

with the understanding that we always start at 1 in the first part of the argument?

**Comment by assignUser** [comment]
Glad you like it! I am not married to the name, wasn't really sure about it... I am new to unit tests too but I think the advantages are clear :D 

I have also found an interesting extension that makes property testing ala haskells's quickcheck possible ( I am currently reading into it and will send a pr with some examples "soon", short form: you describe the spec for the test and the package generates many actual test cases and with functional magic in the background xD)

Yes when you set `link = "fill"` those all do the same, because the reps part of the formula gets overwritten. If fill is not set there will be an error if `length(vector) * reps != n`, the idea behind that was to add a check for the user so that if other stuff in the definition that affects row count changes, this will throw an error. 

For example: each subject is tested in 2 sessions with the same 3 stimuli but in random order so we use `addPeriods(6,data)`. We set  `formula = "4+7+9;12" with `link = "identity"` if the data does not have 36 rows (aka 6 subjects * 2 sessions * 3 stimuli) there will be an error. 

So in case we adjust the experimental design and change the number of stimuli/sessions etc. but forget to adjust the formula we get an error. Maybe that should be more in the hands of the user? The use case for prSeq is still there in my opinion.

I wanted the user to be able to specify the exact elements to use instead of a range, because it is not really more work to parse it and does not create problems (contrary to e.g. unif where we only a range is valid input) and it allows more freedom. I allows for a mix of floats and integers for example. The whole `+` thing is pretty ugly and was mainly to make it past `.devEval` 😁 .

I will mention #12 just so that these comments get linked in the matching PR :)

**Comment by assignUser** [comment]
Sorry if my examples are all pretty specific to my use case, I hope they are still clear none the less.

**Comment by kgoldfeld** [comment]
I think I am starting to understand. It is pretty uncanny that this is what you are trying to do, because I was just doing some simulations that required a similar kind of data generation process, though did not have the requirement that the generated categories be something like your stimuli. 

With my further understanding that you are generating repeated exposures per individual, I don't actually love the data generation process you've created, because it does not reflect the structure of the data, which includes the individual and the different cycles of sequences. I don't like the fact that in the current formulation
  
`formula = "4+7+9;12`

is actually generating sequences for 6 individuals. It is not clear at all.

The process I am showing below is not perfect, but it gives a sense of what I mean for the structure of the data to emanate from the function calls. I wouldn't mind creating a single function that collapses the three calls I make below after creating the individual-level data. Here is the code:

```
dtPat <- genData(6, id = "idPatient")

dtBlock <- addPeriods(dtPat, nPeriods = 2, id = "idPatient", perName = "block", 
                      timeid = "idBlock")

dtSeq <- addPeriods(dtBlock, 3, idvars = c("idBlock"), perName = "seq", 
                       timeid = "idSeq")

dtSeq <- trtAssign(dtSeq, nTrt = 3, balanced = TRUE, strata = "idBlock", 
                      grpName = "pseudoSeq")
```

I just don't think in this case it makes sense to do this as part of the genData function. If I am not making any sense, let me know.

By the way - where are you writing from and what is your research focus?

**Comment by assignUser** [comment]
Interesting, I wasn't aware that `trtAssign` could be used for this. I see your point about the structure, it actually reads better this way, as the was the exposures are assigned is clearer. It is probably due to my background in cs that I tend towards abstraction and generalization, to enable as many use cases as possible (e.g. not being constrained to a sequence of ints ). 

Maybe we can find a compromise where we achieve both, code that emanates the structure of the data while still retaining the tools to solve non-standard use cases we maybe haven't even thought about yet.
In the end it is your package and your choice 😄 which I am happy to abide by.

I am a Masters student in Public Health/Computer Science from Germany.

**Comment by kgoldfeld** [comment]
Public Health/Computer Science - that sounds like an awesome combo.

I was thinking that the function `addPeriods` could actually be modified for the purpose at least creating the structure of blocks or cycles, and `trtAssign` could be modified to accommodate the scalar exposures that you are looking for. Note, there is also `genCluster` which creates repeated observations for a single unit, but those are not necessarily time related. I'm not saying that new functions cannot be created, but it would be nice to build off of the existing structure.

**Comment by assignUser** [comment]
I hope so :D 

Sounds like point to start from, as it would definitely make sense to have this functionality in addPeriods and trtAssign. I haven't looked at the code for those functions yet but maybe we could repurpose (at least for trtAssign).  Let me know how you want to go about that. I am currently looking into the `testhat` extension `hedgehog` that allows for automatically creating tons of test cases that fit a defined form, very promising so far!

Have you thought about the idea I discribed in this issue (see: https://github.com/kgoldfeld/simstudy/issues/14#issuecomment-612637263) ?

**Comment by kgoldfeld** [comment]
I really appreciate all the effort going into `testhat` - I really need to come up to speed. As for proceeding with `addPeriods` and `trtAssign` I am happy for you to give it a go. I just have a general vision, but nothing specific.

As for the intensity issue, I really need to understand a little bit more what the goal is. I am still not getting it. Can you provide me with a clear, simple example?

**Comment by assignUser** [comment]
I think I will focus on the tests for now because that will make all other changes safer and easier.

I'll think on a more general example.

**Comment by kgoldfeld** [comment]
Great - and do you have a name?

________________________________

**Comment by assignUser** [comment]
Hey!

Not sure if you have seen it yet but I may have sneakily prepared .evalDef to work with external variables while doing the refactor for #13 . 

In the formula the variables are prefixed with `..` to distinguish them from `defData` created vars, I choose it like this because data.table uses the same concept.

within the generator functions we just need to add a bit of parsing before the usual `with(eval(`:
``` r

dtSim<- NULL
x3 <- 3
formula <- "7+ ..x3"

# Parsing the external vars
fvars <- all.vars(parse(text = formula))
dotvars <- startsWith(fvars,"..")

ddd <- mget(sub("..", "", fvars[dotvars]),envir = .GlobalEnv)
names(ddd) <- fvars[dotvars]

# Creating an environment from external vars and dtSim.
e <- list2env(ddd)

if(!is.null(dtSim))
  e <- list2env(dtSim,e)

new <- with(e,eval(parse(text = as.character(formula))))
new
#> [1] 10
```

<sup>Created on 2020-04-20 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

This way we don't need to touch the way definition tables work and can still handle more use cases! What do you think?

**Comment by kgoldfeld** [comment]
I think that is pretty awesome. It has always annoyed me that it was not possible to include non-table variables in the formula, which led me to create defFormula, but this is so much better if it can fit in seamlessly without a lot of overhead (i.e. not slow things down). Computing performance (as well as data generation transparency) has always been a guiding idea behind the development of simstudy.

**Comment by assignUser** [comment]
Nice, glad you think so! I just ran a quick benchmark and it looks like the parsing and getting of the vars just takes ~50us (15-20k iterations/second) for what I assume is a relatively long formula:   `nr + ..x2 * y1 ( log(red) * ..x3[me]) / x4 +3`, for a shorter one with just 2 vars it took 40us. 

I don't know if this is within tolerance for you (how big are the data sets you usually create?), but I think we could even speed this up/skip it, because that whole process already happens when checking the formula in .evalDef, so we could pass the parsed names and values back up to the calling functions (defAdd etc.) and down to the generate functions. 

As an aside: I found a [really nifty tutorial](https://www.r-bloggers.com/internal-functions-in-r-packages/) on how to visualize the function calls within a package as a network, it really helped me  understand whats called where on 👍 
![image](https://user-images.githubusercontent.com/16141871/79802672-ae254a80-8360-11ea-8350-a99a757a9f81.png)

**Comment by kgoldfeld** [comment]
The performance is probably fine. It isn't really a matter of how big my data sets are - it is just that I don't want speed to be a limitation for any potential user. And little loss in performance does start to add up when doing 10's of thousands of simulations.


One thing to be mindful of is to make sure sooner rather than later that these types of changes (I'm thinking in particular of creating that temporary environment) might run into trouble when going through the CRAN checks. It might be good to create a test package and do all the tests but not submit. I also know that people have issues with their unit testing. Something to keep in mind.

________________________________

**Comment by assignUser** [comment]
I see, good point. Up till now I skipped the CRAN portion of the r-packages book. Will rectify that.
Also could you elaborate on what you mean by "data generation transparency", so I can conform to your `idea(ls | s)` 😄

**Comment by kgoldfeld** [comment]
I guess when I say data generation transparency I was referring to the idea that hierarchical and relational nature of the model that is being simulated needs to be reflecting in the series of function calls. I don't want a function to be so efficient that it is unclear how it fits into the data generation process. So, I guess the initial example we started with - with the block generation - could be done more efficiently in a less transparent way.


If I have not articulated this very well, no worries. I will know it when I see it, and will point it out to you.

________________________________

**Comment by assignUser** [comment]
Ah yes, I think I get it. 

I think the use of new.env() should not be an issue on CRAN. Long running tests can be though so we will skip those with `skip_on_cran()`. And I will check everything with `R CMD Check`, we could set up travis to check that after every push. 👍

**Comment by kgoldfeld** [comment]
Things have been crazy on my end - been involved with getting 4 Covid-19 related studies up and running. So, I have not been able to dive into all of the work you have been doing. I hope to do that soon, and to figure out how to formally move this forward. In the meantime, continue to keep me posted.

**Comment by assignUser** [comment]
Wow sounds interesting and way more important than this, good luck! We will talk in time. I will finish up writing tests for .evalDef and such, then we can talk about how to go forward 😃

**Comment by assignUser** [comment]
Closed by #22
